### PR TITLE
Display labels instead of icons in top toolbar.

### LINKED ISSUE #10524: Optional Inline Labels in Toolbar
**Is your feature request related to a problem? Please describe.**
Icon only toolbars are fine for some people, and some have a preference for it ( e.g. myself ).

However, some people are unable to effectively use icon only toolbars, perhaps it's their preference, but for some people it has an a11y aspect. This could be similar to face blindness or forms of dyslexia.

For those people tooltips are not enough, and the friction of hovering over each button and waiting for the tooltip to display is immense.

Research indicates that a combination of icons and texts yields the widest and greatest recognition speed possible, at the cost of space

**Describe the solution you'd like**

MacOS solves this problem by providing options when customising toolbars, e.g.:
screenshot 2018-10-11 at 17 24 05

Icons only:

screenshot 2018-10-11 at 17 23 23

Icons and text:

screenshot 2018-10-11 at 17 23 29

Text only:

screenshot 2018-10-11 at 17 23 37

I've mocked up what an icons and text mode may look like in Gutenberg:

screenshot 2018-10-11 at 17 19 59

To me this looks like a clear improvement for some, and makes for much larger target areas on larger devices. I believe the icons only should remain as the default option however

I'm unsure how it would look placed underneath as MacOS does. I'm also unsure how well a text only option would fair, but it would look a lot prettier than MacOS given the current Gutenberg UI styling.

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Fixes #10524 , along with #24234 . This PR points to #24234 so that only the changeset related to displaying button labels in the top toolbar is visible.

All the feedback and comments from #15830 have been addressed. One point that I'd appreciate further feedback on is the positioning of nested dropdowns, such as:

Screen Shot 2020-07-31 at 3 45 49 pm

## How has this been tested?

Tested in browser at different breakpoints. Keyboard navigation checked.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by mapk** [comment]
Thanks for this PR! I spent some time working with the [previous design explorations](https://github.com/WordPress/gutenberg/pull/15830) and wanted to present my further iterations here. While this isn't direct feedback on the PR (yet), I thought it best to share on this PR for a discussion as to what path works best.

----
### Option 1 - Add labels under icons
txt-under-icons

- Reduce the active state of the Settings icon from 36px to 32px. This allows more room for text and aligns it to the same size as the Inserter.
----
### Option 2 - Expand height of topbar
76px

- Height has been expanded from 60px to 76px to allow more room for text.
- The "W", at this size, feels too big to be full bleed. I dropped it down a bit and pulled it away from edges.
- The right side buttons feel a little out of alignment and would need further iterations.
----
### Option 3 - Smaller icons
smaller-icons

**Mobile**

Screen Shot 2020-08-03 at 4 47 00 PM

- I resized the icons from 36px to 28px. This doesn't align with our standard icon size, but does allow for more breathing room when icon labels are visible.
- With the smaller icons, the positioning of the Publish button doesn't feel so bad.
- Again, the Settings icon active size matches the Inserter size.
----
### Option 4 - Text only
Screen Shot 2020-08-03 at 4 48 42 PM

**Mobile**

mobile


- Eliminate all icons (as this PR does already).
- I've explored several iterations with the Mover text.
- As shown in the Mobile view, it would be better to display the block's name in the toolbar rather than just "block".

----

### Feedback
I wanted to get these early iterations up for discussion. Is there an option that feels better than the others? What else would you like to see iterated? If you're interested in more, check out my [Figma file](https://www.figma.com/file/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=0%3A1).

File: packages/components/src/dropdown-menu/index.js
Comment (noisysocks) [inline packages/components/src/dropdown-menu/index.js:132]: This will always set `showTooltip` regardless of what `toggleProps?.showTooltip` is.

Assuming that `true` is supposed to be a default value that can be overridden by `toggleProps?.showTooltip`, we could use `??` instead of `||`:

```suggestion
						showTooltip={ toggleProps?.showTooltip ?? true }
```

**Review by noisysocks — COMMENTED** [review]
Nice job! Seems to work pretty well. Waiting for design to be finalised before I test thoroughly.

File: packages/components/src/dropdown-menu/index.js
Comment (tellthemachines) [reply packages/components/src/dropdown-menu/index.js:132]: Oooh, well spotted! Fixing.

**Comment by mapk** [comment]
Here's my feedback on the current state of this PR:

### Text only
I love the text only mode because it's so clean. But once the Top Toolbar option is introduced, it feels a little wonky. This PR chooses to keep the block toolbar on a second row, which could work, but would need updating to text only as well. (I think that's reserved for another PR)

-----

### Top Toolbar mode
The bottom border is missing in the second level toolbar here:

Screen Shot 2020-08-04 at 4 04 00 PM

-----

### Site icon
I don't think we need to add text here. It just gets very out of place and messy.

Screen Shot 2020-08-04 at 3 58 38 PM

-----

### Text edits
- "Add block" should just be "Add"
- "Block navigation" should just be "List view"

-----

### Settings button
This looks more like a button now than a toggle now. Should keep this pattern?

-----

### Button sizes
When both menu dropdowns are triggered, the "Add block" and "Publish" btns sizes are more visibly unequal and look like a mistake. I think they should all match sizes with this option.

Screen Shot 2020-08-04 at 4 14 29 PM

-----

### Dropdown menus should not overlap
These dropdowns should appear next to their parent level menu.

**Example 1**

menu-drops

**Example 2**

drop2

-----

### Settings item when in dropdown
I think we should drop the "active" state of the Settings item when it's in the dropdown. More like this:

Screen Shot 2020-08-04 at 4 26 29 PM

-----

### Dropdowns
I think we have a standard min-width on dropdowns like 258px. Let's apply that to these as well.

-----
After these changes, I'm happy to provide another review. 😄

**Comment by tellthemachines** [comment]
I've left this in labels-only mode for now, we can always change it later if we decide it doesn't work. Or we could have three options: icons only, icons and labels, and labels only 😄 

I've addressed most of @mapk 's feedback; all but the dropdown positioning as I'm still tussling with the Popover component on that front. Hopefully will have it fixed soon!

The e2e tests are failing because of the changes I made to the label text on some of the buttons. I'd appreciate any feedback on those changes, so that I can fix the tests once they're sorted.

**Comment by mapk** [comment]
Thanks for the quick turnaround, @tellthemachines! As you're working through this, I hope providing a bit more feedback is okay. If you'd like me to hold off until a certain point, just let me know.

### Top Toolbar mode
I noticed that when selecting Top Toolbar mode, when I don't have a block selected the topbar bottom border doubles up. 

border

----

### Text buttons
When switching to a text only mode, I really think all the button interactions should work the same. It always felt odd to me to have icon buttons (ie. hover and pressed states) display differently than the text buttons, and now that these are all text buttons, it's even more obvious. 

**An open question:** Should the text only buttons reflect the same interactions that the Preview button has. These are no longer icon buttons, and should probably now reflect the interactions of our standard button states.

btns

----

### Settings and Options
I just noticed that when both Settings and Options are in a dropdown, and activated, they both show blue. This doesn't feel right to me even though technically it is. Maybe fixing the whole button interaction thing will help this?

Screen Shot 2020-08-05 at 10 46 11 AM

----

### Icon dropdown
In text only mode, when the screen size is reduced, the buttons all move to a dropdown that is represented only with an icon. Does this make sense in the text only mode? Should we have a label for this icon?

Screen Shot 2020-08-05 at 10 57 31 AM


----
There are so many small nuances that come to light as this gets developed, so I really appreciate your time with it all.

**Comment by mapk** [comment]
If we converted all the topbar text-only buttons to the default button instance as outlined in our Figma libraries, they would interact similarly to the Preview button. Check out this prototype below: (please ignore the black background)

btns

**Comment by tellthemachines** [comment]
> As you're working through this, I hope providing a bit more feedback is okay.

Absolutely, all feedback helps!

> I noticed that when selecting Top Toolbar mode, when I don't have a block selected the topbar bottom border doubles up.

Hmm, I can't reproduce this. Have you checked out my latest changes? 

> In text only mode, when the screen size is reduced, the buttons all move to a dropdown that is represented only with an icon. Does this make sense in the text only mode? Should we have a label for this icon?

Very good question 😅 
I'm happy to try out a text label here, just not sure what to call it (we don't currently have an aria-label on this button, so I guess that'll need addressing either way). "More settings" seems a bit weird/redundant? 

I've switched all the text buttons to behave like the Preview button in my latest commit, let me know if it works!

Regarding the dropdown positioning issues, it's trickier than I thought initially. The Popover component that we use in dropdowns has very limited positioning capabilities. We have #21275 to address that, and from the looks of it it's going to require rewriting the whole component. I repositioned the Outline dropdown the best I could, but haven't been able to improve any of the others. Can we leave it as is for now, and revisit once Popover is rebuilt?

**Comment by mapk** [comment]
>I've switched all the text buttons to behave like the Preview button in my latest commit, let me know if it works!

After sitting on this one for a few days, I'm not fond of how our default button state looks for each of these topbar items. I iterated further with something more reminiscent of our Icon Button styles. It's not perfect, but the other direction felt overwhelming.

- The gray pressed state is replaced with a simple 1.5 pixel blue border.
- I aligned with the [recent PR](https://github.com/WordPress/gutenberg/pull/24420) to make the text links in the topbar blue instead of black.
- Hovered states are identified with a border as are the active states.

icon-like-buttons

### Next steps
- Iterate when the Top Toolbar mode is "on". How do the toolbar buttons look when there are gray dividers?
- Explore other interaction designs.

**Comment by mapk** [comment]
I got some feedback yesterday that moving to "text only" might be too much. The text only version loses hierarchy and grouping which can cause usability issues. 

I'm exploring a couple icon + text solutions to see how else we might be able to retain some sense of hierarchy and relationship once we introduce text into the topbar.

cc @kellychoffman

**Comment by ZebulanStanphill** [comment]
@mapk Could you elaborate on how is it losing hierarchy/grouping? I suspect the issue is not with text vs. icons, but with the pre-existing lack of clear division between unrelated controls in the top toolbar. Why aren't there visible dividers like how controls are grouped in the block toolbar? I've said it before, but the top toolbar doesn't seem to have a clear sense of where controls go. (And it's even worse with "Top Toolbar" enabled; the block toolbar should be treated as a separate toolbar below the top one.)

**Comment by tellthemachines** [comment]
> @mapk Could you elaborate on how is it losing hierarchy/grouping?

I'm curious about this too. I'm not aware of any existing hierarchy or grouping (apart from left vs right hand side of the top toolbar), so perhaps that's something that needs to be made clearer in the overall design.

**Comment by mapk** [comment]
@tellthemachines and @ZebulanStanphill You both are correct. While the icons lacked organization, they didn't take up a large amount of space, and the shapes helped differentiate them. The text-only option brings stronger focus to the lack of both their order and their level of hierarchy in relation to each other. Everything is now a similar size and importance. For example, the "redo" and "undo" items feel too prominent.

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (talldan) [inline packages/edit-post/src/components/header/header-toolbar/index.js:154]: I know there was some discussion on the PR about the design of this, but should this also be a text button?

It seems inconsistent to have an icon button in this mode.

It also looks like it needs a label.

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (talldan) [inline packages/edit-post/src/components/header/header-toolbar/index.js:106]: The disabled items in the overflow menu don't seem to be consistent with other dropdown menus:
Screenshot 2020-08-13 at 4 01 11 pm

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (talldan) [inline packages/edit-post/src/components/header/header-toolbar/index.js:85]: More of a question for design, I wonder if it could be worth bringing the tools inline in the overflow menu instead of having them in an additional popover.

So instead of having something like:
Screenshot 2020-08-13 at 4 11 28 pm


It could be more like the Visual Editor / Code Editor options in the more menu in a separate section of the overflow menu:
Screenshot 2020-08-13 at 4 11 45 pm

File: packages/edit-post/src/components/header/index.js
Comment (talldan) [inline packages/edit-post/src/components/header/index.js:96]: Similar again here, this button seems to need a label.

The button styles in the dropdown also look like they need some tweaking:
Screenshot 2020-08-13 at 4 08 04 pm

**Review by talldan — COMMENTED** [review]
Great work on this, I can see it's not at all straightforward to wrangle all that text and make things work on various screen sizes.

Looks pretty close to ready, mostly just the dropdown menus need a bit more work so that they're consistent.

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (ZebulanStanphill) [reply packages/edit-post/src/components/header/header-toolbar/index.js:154]: This raises the question of what label to give this button. There are two of them, and just calling them both "More" seems a bit confusing, since there's no clear answer to "More what?" ...which I guess just loops back to the same issue we keep running into: the top toolbar is not organized... or rather, what organization it does have is very vague and arbitrary.

File: packages/block-editor/src/components/block-navigation/dropdown.js
Comment (tellthemachines) [reply packages/block-editor/src/components/block-navigation/dropdown.js:50]: Ooooh right.

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (talldan) [reply packages/edit-post/src/components/header/header-toolbar/index.js:154]: Hmm, yep, Settings and Options are also often interchangeable, but they are in the same menu at small breakpoints with not much to distinguish them.

I do think there's a bit of a separation in the way options are grouped. The items on the left are more about interacting with content, while those on the right are more related to the editor features or publishing. Not sure I'm smart enough to dilute that into button labels though. The best I can come up with is:
- 'More content tools'
- 'More editor options'

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (tellthemachines) [reply packages/edit-post/src/components/header/header-toolbar/index.js:85]: This relates to the earlier [conversation about hierarchy](https://github.com/WordPress/gutenberg/pull/24304#issuecomment-672964514). 
My only thought from a code standpoint is this will require quite a bit of markup/logic duplication, as we would have to use different components for each breakpoint.

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (tellthemachines) [reply packages/edit-post/src/components/header/header-toolbar/index.js:106]: Yeah, they are meant to be consistent with their out-of-overflow state 😅 but perhaps @mapk can comment further on that.

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (tellthemachines) [reply packages/edit-post/src/components/header/header-toolbar/index.js:147]: I don't think I can use `DropdownMenu` here because its only permitted children are `MenuItem`, `MenuItemsChoice`, or `MenuGroup`, which don't seem right semantically. Unless I'm missing something?

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (tellthemachines) [reply packages/edit-post/src/components/header/header-toolbar/index.js:154]: Yeah, we seem to be going around in circles on this. "More tools and options" has already become "Options", and these dropdowns include that and a heap of other stuff that isn't easy to label as a group 😬

Comment (tellthemachines) [reply packages/edit-post/src/components/header/header-toolbar/index.js:154]: > The best I can come up with is:
> 'More content tools'
> 'More editor options'

That's much better than my 0 ideas so far 😅 
We'd need them to be shorter though. And clicking on an "options" button only to find another "options" button inside would be a bit 😭. 
Putting the sidebar toggle inside the options menu doesn't seem right either. I wonder if we could leave it out and instead pop the "switch to draft" in the "options"?

Comment (ZebulanStanphill) [reply packages/edit-post/src/components/header/header-toolbar/index.js:154]: Perhaps instead of having essentially a more menu inside of another more menu on the right, just have one more menu that stuff moves into as necessary (much like how the overflow menu in the header of the Twenty Nineteen theme works). The toolbar should also be reorganized so the most important controls are listed first, so they remain visible at more viewport widths, e.g. "Publish" should come before "Save draft" and "Preview".

**Comment by tellthemachines** [comment]
I've updated this PR to add text labels to the dropdowns that appear on smaller breakpoints. I had to make a couple of changes to existing labels in order to avoid duplication, so the left hand dropdown is called "Tools" and I changed the existing "Tools" dropdown to be called "Modes" as I think that describes what it does a bit better. The right hand dropdown is now called "Options" and I updated the "More tools & options" button to be called "More options".

This works well on very small breakpoints, even when plugins have registered buttons on the right hand side slot:

Screen Shot 2020-08-25 at 12 02 30 pm

This is something that the current layout with the icon buttons doesn't solve very well, so perhaps we can use this as an exercise to consider further improvements to small breakpoint layouts:

Screen Shot 2020-08-25 at 11 15 22 am

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (mapk) [reply packages/edit-post/src/components/header/header-toolbar/index.js:81]: Separating these under section headings and using dividers could really help. I'll explore some ideas.

**Comment by mapk** [comment]
Taking another look! Here's a visual of the current state of this PR:

current-state

### Design Improvements
There's a couple things I notice right away:
- There are too many various button states that are visible creating a hodge-podge of elements.
- The "undo/redo" items are actually requesting more attention because of their grey background than the simple blue text.
- "Block navigation" and "More options" take up too much space using two words. Making these changes can help:
    - Change "Outline" to "Details"
    - Change "Block navigation" to "Outline"
    - Change "More options" to "Options" (Keep it as "Options" regardless of whether "Settings" is included under it or not)
- "Add block" is too much text too... 
    - Change "Add block" to "Add"

### Desktop menu [prototype](https://www.figma.com/proto/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=234%3A270&viewport=-6704%2C-1092%2C1&scaling=min-zoom)
I haven't revisited the ordering of items in text-only mode yet to address the hierarchy and grouping issues, but this is moving closer to styling improvements on the existing state.

iterated


### Dropdowns
The dropdowns need some iterations as well. I know you're thinking of saving this for another PR, but I thought documenting the changes here would be fine until that PR is created. :)

dropdowns

- The width of the popovers should be `260px` to remain consistent with the other popovers.
- When "Settings" drops into the "Options" dropdown, it loses it's black background state. I think this is fine as long as all buttons in the dropdown are consistent. Right now they aren't (ie. undo/redo).
- The "Options" dropdown no longer sits centered above the Sidebar settings panel. It's moved further to the left. It would be good to keep its positioning.
- All buttons in the dropdowns should have a consistent width of `234px`.
- When menu items get moved into dropdowns, their parent button has a blue border. This is not consistent with the other buttons like "Preview" when in the topbar. Can we remove the border?
- None of the popovers should overlap each other or be displayed below each other as mentioned above.

### Dropdown [prototype](https://www.figma.com/proto/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=476%3A1&viewport=-1545%2C625%2C0.5&scaling=min-zoom)
I iterated on some improvements for the dropdowns/popovers. I would imagine them working more closer to this:

dropdowns

**Comment by tellthemachines** [comment]
I think I've addressed all the feedback now, with the only pending items being the position of "Outline" and "Details" dropdowns. Those are dependent on upcoming Popover component API changes, so should be tackled separately.

@mapk I gave your dropdown prototype rearrangement a go, and it's working pretty well! Let me know what you think 😄

**Comment by mapk** [comment]
Digging in a bit more, I've found a few design issues with the buttons.

### Button padding
The horizontal padding on both the "Add" and "Settings" buttons is too tight.

Screen Shot 2020-08-26 at 4 28 24 PM

and

Screen Shot 2020-08-26 at 4 28 32 PM

### Button proximity
The "Options" button is sitting too close to the "Settings" button. It is visually odd. Can it have the same space between that the "Preview" and "Publish" buttons have?

Screen Shot 2020-08-26 at 4 28 44 PM

**Comment by tellthemachines** [comment]
@mapk all the button styling should be fixed now! 

I think this is ready for another code review too.

**Comment by adamziel** [comment]
Thank you so much for taking on this @tellthemachines - I can see it's a grind, but it's going to help so many people!

A few visual glitches I found:

1. Notice how the buttons stack on iPhone 5. Nothing wrong with that per se, but they're also getting cut off:

Zrzut ekranu 2020-08-27 o 11 02 56
Zrzut ekranu 2020-08-27 o 11 03 04

2. The stacking causes some visual flickering when saving the post:



3. Probably unrelated to this specific PR, but it seems like icon buttons have a tad too large margin on mobile:

Zrzut ekranu 2020-08-27 o 11 04 57

4. Again that's probably unrelated to this specific PR, but I noticed that block toolbar has light background on mobile and transparent background on desktop:

Zrzut ekranu 2020-08-27 o 11 05 08

**Comment by adamziel** [comment]
Also just pointing out - it seems super weird to me to have both "Settings" and "Options" buttons - I have no idea how they're different until I actually click both and then hopefully remember what they do after a few interactions with the editor:

Zrzut ekranu 2020-08-27 o 11 09 41

**Comment by adamziel** [comment]
Let's compare Settings with Options top toolbar buttons:

Zrzut ekranu 2020-08-27 o 11 42 08

It didn't bother me as much when they were just icons, but now that they're text the inconsistencies are apparent. The same applies to tools dropdown.

Speaking of which, in the Options dropdown, all the menu items have black font as on the screenshot above. In the Tools dropdown, there's a mixture of blue and black fonts:

Zrzut ekranu 2020-08-27 o 11 23 02

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (kevin940726) [inline packages/edit-post/src/components/header/header-toolbar/index.js:182]: Nitpick/Question: Do we need to wrap this in a `Fragment`?

File: packages/edit-post/src/components/header/style.scss
Comment (kevin940726) [inline packages/edit-post/src/components/header/style.scss:133]: Is this comment intentional?

**Comment by kevin940726** [comment]
image

Not sure if it's intentional to display the background color as transparent?

image

Also, if we have the "Autosave" banner at the same time, it will overlap with the toolbar.

image

In a smaller viewport, the toolbar is on top of the autosave banner.

**Comment by ZebulanStanphill** [comment]
The ellipsis menu is a mess. We should probably look into cleaning it up in a separate PR. The stuff under "View" are really just editor preferences, so shouldn't they be inside the Options modal? And maybe that options modal should be renamed to "Preferences" or "Editor preferences".

The "Settings" button is interesting. It's vague name is ironically appropriate because it opens a sidebar that combines two things that should never have been combined in the first place: document settings and the block inspector. The only thing they both have in common is that they're both "settings". I think the only real solution here is to split the two apart. Really, the block inspector shouldn't even be accessed from the top toolbar in the first place; it should be accessed from the block toolbar, and it should act like a modal, to better fit accessibility guidelines.

If we can solve problems like these first, then I think it will be a lot easier to land this PR.

**Comment by mapk** [comment]
>Having Options nested in Options does not help either:

Yeah, @adamziel, take a look at everywhere we use the word "Tools" too! 😆  https://github.com/WordPress/gutenberg/issues/19409

I believe a high level of reorganizing can help here. But that's an issue above and beyond this PR.

File: packages/components/src/dropdown-menu/index.js
Comment (tellthemachines) [reply packages/components/src/dropdown-menu/index.js:132]: Yes! Thanks for the suggestion 😄

File: packages/edit-post/src/components/header/header-toolbar/index.js
Comment (tellthemachines) [reply packages/edit-post/src/components/header/header-toolbar/index.js:182]: Whoops, no 😅

File: packages/edit-post/src/components/header/style.scss
Comment (tellthemachines) [reply packages/edit-post/src/components/header/style.scss]: Nope, removing!

**Comment by tellthemachines** [comment]
OK, toolbar transparency and mobile layout issues should be fixed now. I had to change the display text on the "Save as..." button to just "Save" on small breakpoint (mainly because "Save as pending" was breaking the layout no matter how much margin and padding were adjusted).

On "Settings" vs "Options" buttons styling, and on font color differences in the "Tools" dropdown, I will defer to @mapk as I've been following his designs.

Regarding the nested "Options" and the semantics of menu organisation more generally, all the points presented above by @adamziel and @ZebulanStanphill are valid, but solving them will require quite a bit of thinking and reorganising, so they shouldn't block this piece of work. Shipping a V1 of this will force future designs to take text labels into account, so hopefully some of the issues we're encountering here won't happen again 🙂

**Comment by mtias** [comment]
I agree with @tellthemachines — the scope of this PR should be to land the option of having labeled buttons (with or without icon) and any trivial modifications to names that could be considered general improvements. Any further substantial changes to things like the inspector or the options + plugins menu is going to keep punting this improvement.

On the subject of text only and icon + text, I think it might be fine to include the three options (text, icons, text + icons).

Once we land this let's make sure to dedicate some time at reorganizing the settings modal, as it has grown a bit unconcernedly over the past couple months.

**Review by draganescu — APPROVED** [review]
I went through the code and it's 👍 (this PR went through a few code reviews so far) and I tested on browser and mobile. It works acceptable for this phase. The PR is big enough and it implements a change that can be a foundation to build upon.

P.S. I love text only buttons so I would live even the block toolbar to get these labels automatically when it's fixed to top :)

**Review by aristath — APPROVED** [review]
Code looks good to me and it makes sense. Tested and looks great :+1:

**Comment by mapk** [comment]
Caught some spacing issues on smaller screens.

spacing

**Comment by mapk** [comment]
Generally, the spacing in between these buttons all feels a bit off.

**These all feel too close.**

Screen Shot 2020-09-01 at 3 50 10 PM

**These feel too far from one another.**

Screen Shot 2020-09-01 at 3 50 19 PM

Is there a way to make this more consistent across all the buttons in text-only mode? In this Figma mockup, I have them spaced 10px apart.

good-spacing

[Figma mockup](https://www.figma.com/file/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=690%3A52)

**Comment by ZebulanStanphill** [comment]
`10px` isn't aligned to our [grid system](https://github.com/WordPress/gutenberg/blob/master/packages/base-styles/_variables.scss#L27), but yeah, something close to that would make sense.

**Comment by ZebulanStanphill** [comment]
Noting that #24965 has been created to discuss improvements to the "Options" modal, which should help solve some of the overly-similar labels currently present.

See also these other issues relating to the top toolbar:
- #21870 G2: Apply 12px grid to top toolbar more consistently
- #20877 Rethinking the "Current Document" display and settings
- #20592 G2 block top toolbar
- #19409 Two distinct features now labeled "Tools"
- #19128 Exploration: moving options

**Comment by talldan** [comment]
@mapk Was wondering why 'Outline' was used for the button text that opens the List View, instead of 'List View'?

Looking back through the designs in the comments, it looks like at one point it said 'List View' (https://github.com/WordPress/gutenberg/pull/24304#issuecomment-671609218) but then changed in the more recent designs.

**Comment by mapk** [comment]
@ZebulanStanphill, changing the spacing to 12px is totally fine. 

@talldan, No problem changing it to "List View". Let's make that happen. :)

**Comment by afercia** [comment]
Glad to see this merged! Thanks everyone for the perseverance and I'd like to thank @nicolad who started the first attempt to build this in #15830 ❤️ (deserves props).

Regarding this point by @ZebulanStanphill:

> Really, the block inspector shouldn't even be accessed from the top toolbar in the first place; it should be accessed from the block toolbar, and it should act like a modal, to better fit accessibility guidelines.

Ah! That's what the accessibility team has been asking in the last... years? Considering all the troubles for keyboard navigation to go from the selected block to the inspector and back, the teams asked several times to have the inspector "in place" close to the block instead of in the sidebar. Hopefully there will be new explorations in this direction.

**Comment by afercia** [comment]
I'd like to propose props also for @tomjn who created [the original issue](https://github.com/WordPress/gutenberg/issues/10524), together with @nicolad who, as mentioned above, worked hard on the first PR.

**Comment by tomjn** [comment]
I actually raised the issue during a user group after a discussion with @mikelittle, he deserves the props :)

**Pull request merged**
